### PR TITLE
SUS-4169 | Cleanup the database after not used user groups

### PR DESCRIPTION
**Pull request by mszabo-wikia**
Create a maintenance script that will prune not existing / wrongly scoped user group entries from databases. It will be run as part of update.php but can also be executed independently. `--run-on-wikicities` flag can be specified to perform a cleanup of shared DB.

https://wikia-inc.atlassian.net/browse/SUS-4169

### CODE REVIEW

File: extensions/wikia/UserProfilePageV3/tests/UserTagsStrategyTest.php [hunk 1]
```diff
@@ -44,16 +47,18 @@ public function setUp() {
...
 		$permissionsServiceMock->expects( $this->any() )
 			->method( 'getExplicitLocalGroups' )
 			->willReturn( $localGroups );
-		$container = ContainerBuilder::buildDevContainer();
-		$container->set( PermissionsService::class, $permissionsServiceMock );
-		$injector = new Injector( $container );
+
+		$injector = InjectorInitializer::newInjectorWithOverrides( function ( InjectorBuilder $builder ) use ( $permissionsServiceMock ) {
+			$builder->bind( PermissionsService::class )->to( $permissionsServiceMock );
+		} );
```
Comment (macbre) [inline extensions/wikia/UserProfilePageV3/tests/UserTagsStrategyTest.php]: `E_JAVA_WARNING`

File: includes/wikia/WikiaUpdater.php [hunk 2]
```diff
@@ -319,6 +320,19 @@ public static function migrateRecentChangesIpData( DatabaseUpdater $databaseUpda
...
 
+	public static function removeUnusedGroups( DatabaseUpdater $databaseUpdater ) {
+		global $IP;
+
+		$databaseUpdater->output( "Cleaning up after unused user groups...\n" );
+
+		$worker = $databaseUpdater->maintenance->runChild(
+			'RemoveUnusedGroups',
+			"$IP/maintenance/wikia/removeUnusedGroups.php"
+		);
```
Comment (macbre) [inline includes/wikia/WikiaUpdater.php]: Shouldn't we just run a query here instead of starting a new process for each single wiki?

Comment (mszabo-wikia) [reply includes/wikia/WikiaUpdater.php]: name is misleading, it doesn't spawn a new process - it instantiates the class and calls `execute`, so it's all the same

File: extensions/wikia/UserProfilePageV3/tests/UserTagsStrategyTest.php (hunk 1 above)
Comment (mszabo-wikia) [reply extensions/wikia/UserProfilePageV3/tests/UserTagsStrategyTest.php]: ☕️

File: maintenance/wikia/removeUnusedGroups.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	public function execute() {
+		if ( $this->hasOption( 'dry-run' ) ) {
+			$this->output( "Dry-run mode, no changes will be made!\n" );
+
+			$dbr = $this->getDatabase( DB_SLAVE );
+
+			$count = $dbr->selectField(
+				'user_groups',
+				'count(*)',
+			'ug_group NOT IN (' . $dbr->makeList( $this->getValidGroups() ) . ')',
```
Comment (macbre) [inline maintenance/wikia/removeUnusedGroups.php]: Tab missing

File: maintenance/wikia/removeUnusedGroups.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		} else {
+			$dbw = $this->getDatabase( DB_MASTER );
+
+			$dbw->delete(
+				'user_groups',
+				[ 'ug_group NOT IN (' . $dbw->makeList( $this->getValidGroups() ) . ')' ],
+				__METHOD__
+			);
+
+			$this->output( "Deleted {$dbw->affectedRows()} entries referring to nonexistent groups.\n" );
```
Comment (macbre) [inline maintenance/wikia/removeUnusedGroups.php:58]: :+1:

**Review by macbre — APPROVED** [review]

**Review by macbre — APPROVED** [review]

**Pull request merged**